hip/CMakeLists.txt

```diff
@@ -0,0 +1,181 @@
+)
+
+find_package(HIP REQUIRED)
+find_package(hipblas REQUIRED)
+find_package(hipsparse REQUIRED)
+find_path(GINKGO_HIP_THRUST_PATH "thrust/complex.h"
+    PATHS "${HIP_PATH}/../include")
+if (NOT GINKGO_HIP_THRUST_PATH)
+    message(FATAL_ERROR "Could not find the ROCm header thrust/complex.h which is required by Ginkgo HIP.")
+endif()
+
+
+set(GINKGO_HIP_SOURCES
+    base/exception.hip.cpp
+    base/executor.hip.cpp
+    base/version.hip.cpp
+    components/zero_array.hip.cpp
+    factorization/par_ilu_kernels.hip.cpp
+    matrix/coo_kernels.hip.cpp
+    matrix/csr_kernels.hip.cpp
+    matrix/dense_kernels.hip.cpp
+    matrix/ell_kernels.hip.cpp
+    matrix/hybrid_kernels.hip.cpp
+    matrix/sellp_kernels.hip.cpp
+    matrix/sparsity_csr_kernels.hip.cpp
+    preconditioner/jacobi_kernels.hip.cpp
+    solver/bicgstab_kernels.hip.cpp
+    solver/cg_kernels.hip.cpp
+    solver/cgs_kernels.hip.cpp
+    solver/fcg_kernels.hip.cpp
+    solver/gmres_kernels.hip.cpp
+    solver/ir_kernels.hip.cpp
+    solver/lower_trs_kernels.hip.cpp
+    solver/upper_trs_kernels.hip.cpp
+    stop/criterion_kernels.hip.cpp
+    stop/residual_norm_reduction_kernels.hip.cpp)
+
+set(GINKGO_HIPCC_OPTIONS ${GINKGO_HIP_COMPILER_FLAGS})
+set(GINKGO_HIP_NVCC_OPTIONS ${GINKGO_HIP_NVCC_COMPILER_FLAGS})
+set(GINKGO_HIP_HCC_OPTIONS ${GINKGO_HIP_HCC_COMPILER_FLAGS})
+
+set_source_files_properties(${GINKGO_HIP_SOURCES} PROPERTIES HIP_SOURCE_PROPERTY_FORMAT TRUE)
+hip_add_library(ginkgo_hip $<TARGET_OBJECTS:ginkgo_hip_device> ${GINKGO_HIP_SOURCES} HIPCC_OPTIONS ${GINKGO_HIPCC_OPTIONS} NVCC_OPTIONS ${GINKGO_HIP_NVCC_OPTIONS} HCC_OPTIONS ${GINKGO_HIP_HCC_OPTIONS} ${GINKGO_STATIC_OR_SHARED})
+
+if(GINKGO_HIP_AMDGPU AND GINKGO_HIP_PLATFORM MATCHES "hcc")
+    foreach(target ${GINKGO_HIP_AMDGPU})
+        target_compile_options(ginkgo_hip PRIVATE --amdgpu-target=${target})
+        target_link_libraries(ginkgo_hip PRIVATE --amdgpu-target=${target})
+    endforeach()
+endif()
+
+target_compile_options(ginkgo_hip PRIVATE $<$<COMPILE_LANGUAGE:CXX>:${GINKGO_COMPILER_FLAGS}>)
+if(GINKGO_WITH_CLANG_TIDY AND GINKGO_CLANG_TIDY_PATH)
+    set_property(TARGET ginkgo_hip PROPERTY CXX_CLANG_TIDY "${GINKGO_CLANG_TIDY_PATH};-checks=*")
+endif()
+if(GINKGO_WITH_IWYU AND GINKGO_IWYU_PATH)
+    set_property(TARGET ginkgo_hip PROPERTY CXX_INCLUDE_WHAT_YOU_USE ${GINKGO_IWYU_PATH})
+endif()
+
+if(GINKGO_HIP_PLATFORM MATCHES "hcc")
+    # Fix the exception thrown bug with `hcc` backend and shared libraries
+    set_target_properties(ginkgo_hip PROPERTIES LINKER_LANGUAGE HIP)
+
+    # Ban `-hc` flag as INTERFACE_LINK_LIBRARIES since that is propagated when building
+    # a static library, and it's definitely not a known option to any compiler.
+    get_target_property(GINKGO_HCCRT_ILL hcc::hccrt INTERFACE_LINK_LIBRARIES)
+    string(REPLACE "-hc " "" GINKGO_HCCRT_NEW_ILL GINKGO_HCCRT_ILL)
+    set_target_properties(hcc::hccrt PROPERTIES INTERFACE_LINK_LIBRARIES "${GINKGO_HCC_NEW_ILL}")
```

Comment:
should `GINKGO_HCC_NEW_ILL` be `GINKGO_HCCRT_NEW_ILL` ?

Reply:
`string(REPLACE "-hc " "" GINKGO_HCCRT_NEW_ILL GINKGO_HCCRT_ILL)` needs to be `string(REPLACE "-hc " "" GINKGO_HCCRT_NEW_ILL ${GINKGO_HCCRT_ILL})`